airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py

```diff
@@ -469,9 +471,47 @@ def read_records(
         stream_slice: Optional[Mapping[str, Any]] = None,
         stream_state: Optional[Mapping[str, Any]] = None,
     ) -> Iterable[StreamData]:
-        yield from self._read_pages(
-            lambda req, res, state, _slice: self.parse_response(res, stream_slice=_slice, stream_state=state), stream_slice, stream_state
-        )
+        # A cursor_field indicates this is an incremental stream which offers better checkpointing than RFR enabled via the cursor
+        if self.cursor_field or not isinstance(self.get_cursor(), ResumableFullRefreshCursor):
+            yield from self._read_pages(
+                lambda req, res, state, _slice: self.parse_response(res, stream_slice=_slice, stream_state=state),
+                stream_slice,
+                stream_state,
+            )
+        else:
```

Comment:
aren't the two blocks exactly the same?

Reply:
not quite. It's read_page vs read_page**s** .

If the stream has the RFR cursor defined, then we use `self._read_page()` which will only read one page for the invocation of `read_records()`.

Otherwise we use the original `self._read_pages()` to do the full iterating over each page